ingester/ingester.go

```diff
@@ -0,0 +1,699 @@
+
+		ingestedSamples: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prism_ingester_ingested_samples_total",
+			Help: "The total number of samples ingested.",
+		}),
+		discardedSamples: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "prism_ingester_out_of_order_samples_total",
+				Help: "The total number of samples that were discarded because their timestamps were at or before the last received sample for a series.",
+			},
+			[]string{discardReasonLabel},
+		),
+		chunkUtilization: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "prism_ingester_chunk_utilization",
+			Help:    "Distribution of stored chunk utilization.",
+			Buckets: []float64{0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9},
+		}),
+		memoryChunks: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "prism_ingester_memory_chunks",
+			Help: "The total number of samples returned from queries.",
+		}),
+		chunkStoreFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prism_ingester_chunk_store_failures_total",
+			Help: "The total number of errors while storing chunks to the chunk store.",
+		}),
+		queries: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prism_ingester_queries_total",
+			Help: "The total number of queries the ingester has handled.",
+		}),
+		queriedSamples: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prism_ingester_queried_samples_total",
+			Help: "The total number of samples returned from queries.",
+		}),
+	}
+
+	go i.loop()
+	return i, nil
+}
+
+func (i *Ingester) getStateFor(ctx context.Context) (*userState, error) {
+	userID, err := user.GetID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("no user id")
+	}
+
+	i.userStateLock.Lock()
+	defer i.userStateLock.Unlock()
+	state, ok := i.userState[userID]
+	if !ok {
+		state = &userState{
+			userID:     userID,
+			fpToSeries: newSeriesMap(),
+			fpLocker:   newFingerprintLocker(16),
+			index:      newInvertedIndex(),
+		}
+		state.mapper = newFPMapper(state.fpToSeries)
+		i.userState[userID] = state
+	}
+	return state, nil
+}
+
+// NeedsThrottling implements storage.SampleAppender.
+func (*Ingester) NeedsThrottling(_ context.Context) bool {
+	return false
+}
+
+// Append implements storage.SampleAppender.
+func (i *Ingester) Append(ctx context.Context, samples []*model.Sample) error {
+	for _, sample := range samples {
+		if err := i.append(ctx, sample); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (i *Ingester) append(ctx context.Context, sample *model.Sample) error {
+	for ln, lv := range sample.Metric {
+		if len(lv) == 0 {
+			delete(sample.Metric, ln)
+		}
+	}
+
+	i.stopLock.RLock()
+	defer i.stopLock.RUnlock()
+	if i.stopped {
+		return fmt.Errorf("ingester stopping")
+	}
+
+	state, err := i.getStateFor(ctx)
+	if err != nil {
+		return err
+	}
+
+	fp, series, err := state.getOrCreateSeries(sample.Metric)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		state.fpLocker.Unlock(fp)
```

Comment:
Oh I see, `getOrCreateSeries` does. I think that's probably unusual enough that a comment would be helpful.